lib/std/http/Client.zig

```diff
@@ -0,0 +1,142 @@
+    }
+
+    pub fn end(req: *Request) !void {
+        req.headers.appendSliceAssumeCapacity("\r\n");
+        switch (req.protocol) {
+            .http => {
+                try req.stream.writeAll(req.headers.items);
+            },
+            .https => {
+                try req.tls_client.writeAll(req.stream, req.headers.items);
+            },
+        }
+    }
+
+    pub fn read(req: *Request, buffer: []u8) !usize {
+        switch (req.protocol) {
+            .http => return req.stream.read(buffer),
+            .https => return req.tls_client.read(req.stream, buffer),
+        }
+    }
+
+    pub fn readAll(req: *Request, buffer: []u8) !usize {
+        return readAtLeast(req, buffer, buffer.len);
+    }
+
+    pub fn readAtLeast(req: *Request, buffer: []u8, len: usize) !usize {
+        var index: usize = 0;
+        while (index < len) {
+            const amt = try req.read(buffer[index..]);
+            if (amt == 0) {
+                switch (req.protocol) {
+                    .http => break,
+                    .https => if (req.tls_client.eof) break,
+                }
+            }
+            index += amt;
+        }
+        return index;
+    }
+};
+
+pub fn deinit(client: *Client) void {
+    assert(client.active_requests == 0);
+    client.headers.deinit(client.allocator);
+    client.* = undefined;
+}
+
+pub fn request(client: *Client, options: Request.Options) !Request {
+    var req: Request = .{
+        .client = client,
+        .stream = try net.tcpConnectToHost(client.allocator, options.host, options.port),
+        .protocol = options.protocol,
+        .tls_client = undefined,
+    };
+    client.active_requests += 1;
+    errdefer req.deinit();
+
+    switch (options.protocol) {
+        .http => {},
+        .https => {
+            req.tls_client = try std.crypto.tls.Client.init(req.stream, client.ca_bundle, options.host);
+        },
+    }
+
+    try req.headers.ensureUnusedCapacity(
+        client.allocator,
+        @tagName(options.method).len +
+            1 +
+            options.path.len +
+            " HTTP/1.1\r\nHost: ".len +
+            options.host.len +
+            "\r\nUpgrade-Insecure-Requests: 1\r\n".len +
+            client.headers.items.len +
+            2, // for the \r\n at the end of headers
+    );
+    req.headers.appendSliceAssumeCapacity(@tagName(options.method));
+    req.headers.appendSliceAssumeCapacity(" ");
+    req.headers.appendSliceAssumeCapacity(options.path);
+    req.headers.appendSliceAssumeCapacity(" HTTP/1.1\r\nHost: ");
+    req.headers.appendSliceAssumeCapacity(options.host);
+    switch (options.protocol) {
+        .https => req.headers.appendSliceAssumeCapacity("\r\nUpgrade-Insecure-Requests: 1\r\n"),
+        .http => req.headers.appendSliceAssumeCapacity("\r\n"),
+    }
+    req.headers.appendSliceAssumeCapacity(client.headers.items);
+
+    return req;
+}
+
+pub fn addHeader(client: *Client, name: []const u8, value: []const u8) !void {
```

Comment:
note

The fact that you don't bother concatenating headers with the same name avoids having an exception for `Set-Cookie`.